http_server/string_contains_match.py

```diff
@@ -0,0 +1,96 @@
+from amaranth import Module, Signal, Array
+from amaranth.lib.wiring import In, Out, Component
+from amaranth.lib import stream
+from capitalizer import Capitalizer
+
+
+class StringContainsMatch(Component):
+    """
+    Matches a literal string in a stream.
+
+    Looks to see if the stream contains the substring, even if it doesn't start
+    with it. Equivalent to the regex ".*{message}"
+
+    Returns "accepted" upon match.
+
+    Parameters
+    ----------
+    message:    str
+                String to match.
+    match_case: bool, default true
+                Perform a case-sensitive match.
+                Case folding is ASCII-only.
+
+    Attributes
+    ----------
+    input:      Stream(8), in
+                Data stream to match.
+    accepted:   Signal(1), out
+                High if the string has been matched.
+    reset:      Signal(1), in
+                Reset and await new input.
+    """
+
+    input: In(stream.Signature(8))
+    accepted: Out(1)
+    reset: In(1)
+
+    def __init__(self, message: str, match_case: bool = True, **kwargs):
+        super().__init__(**kwargs)
+
+        cased_message = message
+        if not match_case:
+            cased_message = message.upper()
+        self._message = Array(map(ord, cased_message))
+
+        self._match_case = match_case
+
+
+    def elaborate(self, _platform):
+        m = Module()
+
+        # Case-normalized data:
+        c = Signal(8)
+        if self._match_case:
+            m.d.comb += c.eq(self.input.payload)
+        else:
+            m.submodules.capitalizer = capitalizer = Capitalizer()
+            m.d.comb += [
+                    capitalizer.input.eq(self.input.payload),
+                    c.eq(capitalizer.output),
+            ]
+
+        import math
+        size = math.ceil(math.log2(len(self._message))) + 1
+        idx = Signal(size)
+
+        with m.FSM():
+            with m.State("matching"):
+                m.d.comb += [
+                    self.input.ready.eq(1),
+                    self.accepted.eq(0),
+                ]
+                m.next = "matching"
+                with m.If(self.input.valid & ~self.reset):
+                    # Consume one byte; we're already holding "ready".
+                    with m.If(c == self._message[idx]):
+                        with m.If(idx < len(self._message) - 1):
+                            # Accept this character and move on.
+                            m.d.sync += idx.eq(idx + 1)
+                        with m.Else():
+                            # At the end of the string.
+                            m.next = "accepted"
+                    with m.Else():
+                        # Restart search
+                        m.d.sync += idx.eq(0)
```

Comment:
I think this means we aren't matching `/.*{message}/`?

Consider `/.*ididx/`; it should match the string `idididx`. However, here we'll make it to `idid_`, match `i` instead of `x`...and restart the search index, failing to match. (Am I understanding this module correctly?)

Thinking through the problem... for a message of length `L`, if we have `L` bits in a shift register, each tracking "do we match as of this byte- that would get us `.*message`. Or maybe we don't need that for this case, and this kind of matching is sufficient for this problem -- we just shouldn't call it equivalent to `.*message`. Thoughts?

Reply:
Yeah, tested: https://github.com/cceckman/http-accel/blob/128208250c50dc0eef92c7e6d4f4c4654914e67a/http_server/string_contains_match_test.py#L30

Reply:
I think this kind of matching is sufficient for the application ("match \r\n\r\n at end-of-headers),  but... turning it into a shift register match was kind of fun.  

Done.